### DISCUSSION TITLE
Can you have a declaratively-defined table And a flask_sqlalchemy.SQLAlchemy().Table description of the same table in the same program?

### DISCUSSION THREAD

**Discussion by dstromberg** [sqlalchemy/sqlalchemy#9325] [Usage Questions] [answered]
Hi folks.

I have a table defined like:
```py
PipelineVariantTable = db.Table('pipeline_variants', db.metadata,
  db.Column('pipeline_id', db.Integer, db.ForeignKey('pipelines.id',
      name='fk_pipeline_variants_pipeline_id',
      onupdate='CASCADE',
      ondelete='RESTRICT'),
    primary_key=True),
...
```

...and it seems to work fine as a secondary on another table.

But when I try to ORM-compose a query using it, I get:
```
  File "/data/home/dstromberg/src/publishing-jiras/BGS-30756-AutoCleanup-is-unpublishing-live-builds-on-Cortez-Live-Vendor/game-publishing/services/api/./sa-orm-query-2023-02-10", line 110, in main
    pv.pipeline_id == p.id,
AttributeError: 'Alias' object has no attribute 'pipeline_id'
```

Right now, that query looks like:
```py
    pv = aliased(PipelineVariantTable, name='pv')

    lv = rick_subquery(session, v, nv, p)
    overall_query = session.query(nv.id, func.min(bs.build_id)) \
        .join(lv) \
        .where(and_(
            v.id == lv.c.id,
            v.variant_id == nv.id,
            bs.version_id == v.id,
            br.id == bs.branch_id,
            pv.pipeline_id == p.id,
            pv.variant_id == nv.id,
        )) \
        .group_by(nv.id) \
        .order_by(nv.id)
```

But that's what's giving the traceback above.

Most of the tables in this program are set up declaratively, and I'm thinking maybe PipelineVariantTable should be set up declaratively (too?) in order to enable using it in queries.

My question is, if I leave the preexisting PipelineVariantTable as it is for use in secondary relationships, and create a new PipelineVariant declaratively, could that cause problems somewhere down the road?  My biggest concern is that it might cause the same table to be used two ways in the same program, and each be cached separately.

I'm using:
```
$ python3 -m pip list -v | grep -i sqlalchemy
Flask-SQLAlchemy       2.5.1         /data/home/dstromberg/.local/lib/python3.10/site-packages pip
SQLAlchemy             1.4.36        /data/home/dstromberg/.local/lib/python3.10/site-packages pip

$ python3 -m pip list -v | grep -i mysql
mysqlclient            2.1.1         /data/home/dstromberg/.local/lib/python3.10/site-packages pip
PyMySQL                0.8.0         /data/home/dstromberg/.local/lib/python3.10/site-packages pip

bash-4.2# mysql --version
mysql  Ver 14.14 Distrib 5.7.41, for Linux (x86_64) using  EditLine wrapper

```
Suggestions please?

Thanks!

**Comment by CaselIT** [comment, marked as answer]
Hi,

A table is not an ORM model, so what you are doing is equivalent to `pv = PipelineVariantTable.alias('pv')`.

At that point `pv` behaves just like a Table, so you access columns using the `.c` collection, so `pv.c.pipeline_id` etc

**Comment by CaselIT** [comment]
somewhat related to this, should `aliased(table)` raise an error (or start raising deprecation warnings)? cc @zzzeek

**Reply by zzzeek** [reply]
aliased(table) returns an Alias object, that's intentional

**Reply by CaselIT** [reply]
Interesting, so `aliased(table)` is a synonym of `table.alias()`. I though it was a side effect and not suggested to do `aliased(not_a_model)`

**Reply by zzzeek** [reply]
yeah it uses inspect() and then decides what kind of alias to make, either AliasedClass or Alias.   There should probably be more of an ``__aliased__()`` method internally but it doesnt matter much right now.
